### PR TITLE
Arcade2D Module

### PR DESCRIPTION
**Pull request by tituschewxj**
# Description

Adds a new module called Arcade2D that supports animation of 2D drawings, sprite images, keyboard and mouse input, and audio output. The intended users of this module are CS1101S students who have completed up to Source 3. Arcade2D is a wrapper for the Phaser 3, and aims to simplify and streamline the implementation and allow students to make games using code similar to that they have already learnt.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

We tested this by creating a demo game, and other examples and the features work as expected.
Some examples are documented in the index file.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: scripts/src/templates/module.ts
Comment (larrywang0701) [inline scripts/src/templates/module.ts:1]: It seems that your fork or local development enviroment is not synced properly with the main repo. You are trying to commit to these files related to the overall module building system, and I don't think an individual module should modify these files without a very well-described and detailed reason. Try not commit to these files if you not intend to do so.

File: src/bundles/arcade_2d/index.ts
Comment (larrywang0701) [inline src/bundles/arcade_2d/index.ts:1]: Personally I think better with more examples? Since you know your module and every functions inside very well, but for new users this may not. A concept that may looks very simple and basic to you, but may be very unfamiliar to people who just learn game programming, or using your module. As the author of Unity Academy module, I can understand what's "main loop" and "game objects" are, but for new programmers, they may even don't know about the fact that "video games is a big loop". Same thing happened to my Unity Academy module before when my friends are trying to understand my documentation. So maybe try add more examples and documentations to help others better understand your module?

**Review by larrywang0701 — CHANGES_REQUESTED** [review]
Please try not touch the module build system files, or please provide a very detailed reason on why you need to modify the overall module build system files. If you can't 100% know what you are doing, this may cause errors and side effects in other modules.

**Review by larrywang0701 — COMMENTED** [review]
Please confirm on that is it necessarily to change the overall shared module build system files, and please self-check on the security and performance stuffs (for a game engine I think it would lead to a waste of resources (CPU, GPU, battery, etc) on user's devices if it just simply always run in background until user totally close the browser page. In Unity Academy users can choose to manually terminate the engine, and Unity Academy has auto sleep design to reduce resource cost. But since pixnfilx don't have such design, I'm not really sure about the situation in your module here.

File: scripts/src/templates/utilities.ts
Comment (tituschewxj) [reply scripts/src/templates/utilities.ts:1]: Because when I tried to name it without the edit I get the warning below. I'm not sure how you were able to name it with it being valid. I don't think this will affect other modules, since the function `isSnakeCase` is called only when a module is created and this change only expands the possible set of bundle names. I have tested the yarn scripts on this module and doesn't seem to be any problem arising from this change in definition for snake case in the module building process.
Screenshot 2023-04-12 211144

Comment (larrywang0701) [reply scripts/src/templates/utilities.ts:1]: Try directly name it in the json file and path. You don't need to run `yarn create` again after you already created it. Actually I didn't use the `yarn create` command to create the module, and you can also see that the name `physics2d` is also a valid name without changing the file (refer to their PR). Personally I think this modification would unnecessary, if you think this is OK, I can approve it but remember please always be responsible to your code.

Comment (tituschewxj) [reply scripts/src/templates/utilities.ts:1]: Directly renaming the module is an option. However, I feel it is inconsistent if there are module names that violate the snake case naming convention. Suppose there are other modules that want to include numbers in their names in future (following our modules), they might find the same problem of being unable to do so unless they directly rename (which defeats the purpose of not modifying the snake case definition and gives extra work to the creators of that new module)

Comment (larrywang0701) [reply scripts/src/templates/utilities.ts:1]: OK, I think you do have a point, and just allowing numbers in module name seems has no effect on the current modules. I'll approve for you now, and after you self-check on the security stuffs I mentioned (especially on the part that students need to provide an arbitrary string or a function object as a parameter, like your `create_sprite` or `create_text` functions.) and confirmed that all things are ok, you can merge it by yourself. Please remember always be responsible to your code and the users when you do that. Thanks :)

(An injection could happen when attackers intentionally put some JS code as strings into your vulnerble functions' parametrs to let the VM interpret it and execute as JS code, and for most freshman in CS1101S I believe they won't check the code from Source Academy's share links before they press Run button)

Also, a small additional advice here, you may consider this in the future to better help users to use your module. Maybe you can consider adding a parameter type check using the javascript keywork typeof. For example, what will happen if students provide wrong parameter types like `create_circle("123")`? A small improvement could be done to help students use the module better :)

**Review by larrywang0701 — APPROVED** [review]
You can merge it by yourself if you finished self-checking and think all things are safe and well.

File: src/bundles/arcade_2d/functions.ts
Comment (RichDom2185) [inline src/bundles/arcade_2d/functions.ts:75]: Minor nitpick: perhaps, instead of leaving a bunch of "naked" global variables, it would be better to encapsulate this information in some sort of settings object.

File: src/tabs/ArcadeTwod/index.tsx
Comment (RichDom2185) [inline src/tabs/ArcadeTwod/index.tsx:57]: Boolean variables should ideally have names that show they are boolean for readability.

```suggestion
      isPaused: false,
```

**Review by RichDom2185 — CHANGES_REQUESTED** [review]
Hi, great work on the module! Was looking through some of the PRs and leaving my comments as below, do take a look at them at let me know what you think!

Thanks for the hard work!

File: src/bundles/arcade_2d/types.ts
Comment (RichDom2185) [reply src/bundles/arcade_2d/types.ts:27]: Same thing for the `[boolean, boolean]`.

Comment (xenosf) [reply src/bundles/arcade_2d/types.ts:27]: Thank you for the suggestion. I've changed the code to use type aliases.

File: src/bundles/arcade_2d/functions.ts
Comment (tituschewxj) [reply src/bundles/arcade_2d/functions.ts]: Ok, thanks for the suggestion. I have encapsulated the information in a config object.

File: src/bundles/arcade_2d/phaserScene.ts
Comment (RichDom2185) [inline src/bundles/arcade_2d/phaserScene.ts:56]: I'm not sure why the type cast here is necessary. Did you mean:

```suggestion
  private phaserGameObjects: Array<PhaserGameObject> = [];
```

File: src/bundles/arcade_2d/gameobject.ts
Comment (RichDom2185) [inline src/bundles/arcade_2d/gameobject.ts:105]: Perhaps this should be using the `ColorRGBA` type alias instead?

File: src/bundles/arcade_2d/audio.ts
Comment (RichDom2185) [inline src/bundles/arcade_2d/audio.ts:13]:
```suggestion
  // Stores all the created AudioClips
```

**Review by RichDom2185 — COMMENTED** [review]
Thanks for working on the changes! There are some more minor comments as below. No rush to do them, please focus on your exams first.

A not-so-minor nitpick, while it's good that the boolean state variable `paused` has been changed to `isPaused`, there are still quite a few more booleans/functions that return booleans that are not following the usual naming convention of `isXXX`, `hasXXX`, `shouldXXX`, etc., and this affects readability.

For example, under `audio.ts`, possible changes may be (feel free to change accordingly if you feel that a better name captures the purpose of the variable more effectively):

* `loop` &rarr; `shouldLoopAudio`
* `getLoop()` &rarr; `isLooping()`
* `playClip` &rarr; ...
* `loopAudioClip()` &rarr; `setShouldLoopAudio()`
* ...

Note the naming of the setters (generally starting with "set").

Thanks for the hard work!

**Comment by martin-henz** [comment]
What's the status of this PR?

**Comment by tituschewxj** [comment]
The status of this PR is that there are some minor changes to do from the reviewers' comments. We should be ready to merge once we have resolved the reviewers' comments and suggestions.

File: src/bundles/arcade_2d/gameobject.ts
Comment (tituschewxj) [reply src/bundles/arcade_2d/gameobject.ts:105]: Thanks for all the help!

File: src/bundles/arcade_2d/audio.ts
Comment (RichDom2185) [inline src/bundles/arcade_2d/audio.ts]: Minor nit. Since this is not exactly a "temporary" variable (the value is returned, after all), perhaps it would make more sense to call this `prevValue`?

```suggestion
    const prevValue = !this.isUpdated;
    this.setAudioClipUpdated();
    return prevValue;
```

**Review by RichDom2185 — COMMENTED** [review]
Just one minor nitpick. Great work guys! I'll leave it to the other reviewers for further comments.

**Review by larrywang0701 — APPROVED** [review]
If there's no more problems in this PR, then I think it can be merged. You can merge it by yourself if you think everything is ok and ready

**Review by larrywang0701 — APPROVED** [review]

**Review by RichDom2185 — APPROVED** [review]
Thanks, LGTM! Feel free to post some sample programs (via a shared link) to demonstrate these new additions :)

Thanks for the hard work!

**Pull request merged**
